### ISSUE TITLE
[AMD][kernel][Tier 5] MX-FP4 grouped GEMM for MoE expert dispatch on MI355X

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/aiter#4]
## Goal

Add an **MX-FP4 group-GEMM kernel** to AITER for MoE expert dispatch on
MI355X (gfx950). The existing AITER FP8 group-GEMM (`gemm_a8w8_blockscale`)
is the hot path for MoE decode, but MX-FP4 weights up-cast to FP8 inside
the kernel, throwing away MI355X's native MX-FP4 XDLOPS throughput.

A group-GEMM dispatch arm that consumes MX-FP4 scales directly is the
missing primitive for the wider sglang/vLLM MX-FP4 MoE story (cf.
amdpilot-org/sglang #25, #26, #30).

Reuses cached **`amd/Qwen3-235B-A22B-Instruct-2507-MXFP4`** (MoE,
128 experts) and **`amd/Kimi-K2.5-MXFP4`** (MoE, 384 experts).
No new download.

## Hardware / workload

```
Hardware:    8 × AMD MI355X (single node, gfx950), TP=8
Model:       amd/Qwen3-235B-A22B-Instruct-2507-MXFP4  (cached)
Workload:    decode batch=128 tokens, top-k=8 expert dispatch
             (per-token), seqlen=4096
Baseline:    AITER gemm_a8w8_blockscale (FP8) on the same shapes,
             with an MX-FP4 → FP8 dequant prelude
Target:      decode tokens/s strictly higher than the FP8 baseline at
             the same accuracy (max_abs_diff < 5e-3 on logits)
```

## Why MI355X

gfx950 exposes MX-FP4 dot-product via XDLOPS at one MX block / cycle
when shapes line up. The per-expert FFN GEMMs (M=tokens_per_expert,
N=intermediate, K=hidden) hit the sweet spot — for Qwen3-235B's
`intermediate_size=4096`, `hidden=8192`, the K dimension is 4× a
single XDLOP block, so the MX-FP4 win is largest. MI300X lacks the
MX-FP4 lane → this is gfx950-specific; route via `gpu_arch=mi355x`.

## Scope

1. Implement `gemm_a4w4_grouped` in `aiter/ops/triton/_triton_kernels/`
   following the existing `gemm_a8w8_blockscale_grouped` layout
   (group-32 quant blocks, per-block scales).
2. Wire the new kernel into the AITER Python entry point
   (`python/aiter/ops/__init__.py`) under a new dispatch arm gated on
   `weight_dtype == "mxfp4"` AND `gpu_arch == "gfx950"`. Falls back to
   the FP8 path on MI300X.
3. Add a unit test for numerical parity vs the FP8-path reference at
   `groups=8 M=128 N=4096 K=8192` (max_abs_diff < 1e-2, mean_abs_diff
   < 1e-3).
4. Add a benchmark harness `tests/op_tests/test_gemm_a4w4_grouped.py`
   that emits `decode_tokens_per_second` for `B=128, S=4096`.

## Acceptance

- Unit test passes.
- Benchmark on a single MI355X node reports
  `decode_tokens_per_second > 1.0 × baseline_FP8_tokens_per_second`
  (any positive speedup; the FP8 baseline already includes the dequant
  prelude, so even break-even with no extra mem traffic is a win).
- Accuracy gate: max_abs_diff < 5e-3 vs FP8 reference on first-token
  logits for a single sequence.

**Comment by jhinpan** [comment]
**AMDPilot — accumulated progress on this issue**

_Auto-generated from successive trials. Latest update: 2026-05-23T17:18:49Z by `issue-aiter-4-r19`._

### Verified infra fixes (apply during Stage 0)

- Confirmed gemm_a8w8_blockscale_bpreshuffle_asm imports cleanly, isolating crash to kernel execution logic
- Validated request_harness_patch tool availability at /usr/local/bin/request_harness_patch
- Applied StrReplaceFile edit to gfx950_a8w8_blockscale_asm dispatcher to fix x_scale layout hypothesis
- Increased step timeout from 120s to 180s to eliminate SIGKILL deadlocks
- Confirmed moe_gemm_a4w4 kernel functions end-to-end with PYTHONPATH=/workspace/aiter, achieving 1743.95 iters/s


### Working approaches and their best metric

- Applied StrReplaceFile edit to gfx950_a8w8_blockscale_asm dispatcher to address x_scale layout hypothesis
- Committed 162 lines establishing grouped dispatch path for MX-FP4 A4W4 on gfx950


### Approaches that did NOT help — don't retry these

- Direct execution of bench_dispatch_combine.py failed due to missing tests.python.ops module in PYTHONPATH
- Hallucinated non-existent base image 'amdpilot-stage0/issue-aiter-4-r3:stage0' causing immediate bundle rejection
- Scheduler assigned job to cpu-only node despite task requiring GPU for MXFP4 models
- Agent hallucinated non-existent base image 'amdpilot-stage0/issue-aiter-4-r4:stage0' after fixing r3 reference
- Job assigned to cpu-only node 'smci355-ccs-aus-n08-13' despite requiring GPU for MXFP4 models
- Agent ignored nudge to run bench_mxfp4.py, proceeding to commit unbenchmarked code
- Rewriting probe scripts without intermediate validation of single-file version
- Read-only exploration loop persisted for 18 steps (~45 mins) despite 'MANDATORY PROFILING' nudge, delaying first code mutation.
- Searching filesystem for .whl files after harness failure instead of fixing PYTHONPATH or mounting source
- Attempting to remove GPU acquisition hard-fail in harness amendment (rejected by supervisor)
- Re-executing model_download logic post-Stage0 without adapting to known broken symlink state
- Granular file reading (steps 5-8) without synthesizing a modification plan or draft patch


### Open ideas to try next

- Surface 'hallucinated base' errors as primary INFRA bullet immediately upon detection
- Inject list of valid amdpilot-stage0 image tags into Stage 0 formulation prompt context
- Provide list_available_stage0_images() tool to prevent base image hallucination
- Implement pre-flight check in Stage0Formulator to validate symlink targets before model_download
- Fix harness_metric_block_missing error in test_harness.py to satisfy cycle-75 contract
- Implement step-count throttle in ExecutionMonitor to force WriteFile/Shell action after >5 consecutive read-only ops post-nudge.
- Debug prof_fp8_grouped.py logic to ensure it triggers the specific kernel path required to emit the AMDPILOT_METRIC block.
- Modify DockerManager.start_container to mount built amdpilot wheel/source and prepend to PYTHONPATH
- Run python3 /workspace/test_harness.py immediately after kernel success to validate environment
- Repair broken symlink at /it-share-2/data/amdpilot-cache/hf/hub/models--amd--Qwen3-235B-A22B-Instruct-2507-MXFP4
- Add AMDPILOT_METRIC block to test_harness.py to satisfy cycle-75 contract
- Synthesize modification plan for grouped dispatch based on mapped moe_op vs gemm_op topology


_amdpilot bot. Edit between the markers to add/correct fixes; the next update will preserve your changes. To disable updates on this issue, replace the body with the marker line only._

**Comment by sunxxuns** [comment]
## Closing as not tractable by the current autonomous agent (2026-05-23)

This issue has accumulated 6+ chronic-fail retries with 0 wins over the past 12 h on
the autonomous loop. Triage: MX-FP4 grouped GEMM — needs AMD MFMA expertise the agent lacks.

The agent (Kimi-K2.6) does not have sufficient AMD gfx950/gfx942 MFMA + LDS-bank-conflict
expertise to make iterative progress on this scope autonomously. Every retry is
landing on the same blocker (lack of arch-specific reasoning), wasting fleet GPU time.

**To re-open**:
- Pre-decompose into one of: "implement a tile-shape sweep harness" (which the agent CAN do),
  or "verify numerical parity with an upstream reference at single-shape S" (small + provable).
- OR escalate to a human kernel engineer; the broad multi-kernel scope here is not autonomous-friendly.

Closing this round so the fleet doesn't keep burning retry cycles. The work itself
is still valuable — the wrapping just needs to be different.
